sources/api/early-boot-config/src/compression.rs

```diff
@@ -0,0 +1,222 @@
+//! This module supports reading from an input source that could be compressed or plain text.
+//!
+//! Currently gzip compression is supported.
+
+use flate2::read::GzDecoder;
+use std::fs::File;
+use std::io::{BufReader, Chain, Cursor, ErrorKind, Read, Result, Take};
+use std::path::Path;
+
+/// "File magic" that indicates file type is stored in a few bytes at the start at the start of the
+/// data.  For now we only need two bytes for gzip, but if adding new formats, we'd need to read
+/// more.  (The simplest approach may be to read the max length for any format we need and compare
+/// the appropriate prefix length.)
+/// https://en.wikipedia.org/wiki/List_of_file_signatures
+const MAGIC_LEN: usize = 2;
+
+// We currently only support gzip, but it shouldn't be hard to add more.
+/// These bytes are at the start of any gzip-compressed data.
+const GZ_MAGIC: [u8; 2] = [0x1f, 0x8b];
+
+/// This helper takes a slice of bytes representing UTF-8 text, which can optionally be
+/// compressed, and returns an uncompressed string.
+pub fn expand_slice_maybe(input: &[u8]) -> Result<String> {
+    let mut output = String::new();
+    let mut reader = OptionalCompressionReader::new(Cursor::new(input));
+    reader.read_to_string(&mut output)?;
+    Ok(output)
+}
+
+/// This helper takes the path to a file containing UTF-8 text, which can optionally be compressed,
+/// and returns an uncompressed string of all its contents.  File reads are done through BufReader.
+pub fn expand_file_maybe<P>(path: P) -> Result<String>
+where
+    P: AsRef<Path>,
+{
+    let path = path.as_ref();
+    let file = File::open(&path)?;
+    let mut output = String::new();
+    let mut reader = OptionalCompressionReader::new(BufReader::new(file));
+    reader.read_to_string(&mut output)?;
+    Ok(output)
+}
+
+// =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
+
+/// This type lets you wrap a `Read` whose data may or may not be compressed, and its `read()`
+/// calls will uncompress the data if needed.
+pub enum OptionalCompressionReader<R> {
```

Comment:
I suppose this weird enum could be hidden?

```suggestion
pub struct OptionalCompressionReader<R>(WeirdEnum);

pub enum WeirdEnum<R> {
```